Set WM_CLASS to given title on supported platforms

Tezkerek opened the pull request:
This fixes #1350. 

I don't know how elegant this is, if you have suggestions I'd be glad to apply them.

chrisduerr commented on src/window.rs:
```diff
@@ -322,13 +322,13 @@ impl Window {
 
     #[cfg(not(any(target_os = "linux", target_os = "freebsd", target_os = "dragonfly", target_os = "openbsd")))]
-    fn platform_builder_ext(window_builder: WindowBuilder) -> WindowBuilder {
+    fn platform_builder_ext(window_builder: WindowBuilder, wm_class: &str) -> WindowBuilder {
```
This will probably read to a warning on macos. To prevent that it's probably a better idea to change this to `_: &str` instead of `wm_class: &str`.

Tezkerek commented on src/window.rs:
```diff
@@ -209,7 +209,7 @@ impl Window {
             .with_transparency(true)
             .with_decorations(window_config.decorations());
-        let window_builder = Window::platform_builder_ext(window_builder);
+        let window_builder = Window::platform_builder_ext(window_builder, title);
```
Well, at least on Linux, the `--title` flag seems pretty useless, since running a command will change the title to the process name anyway.

nixpulvis replied:
Agreed, maybe the `--title` flag should disable the program name title updates. But either way I think the behavior of `--class` is static.

chrisduerr replied:
The `--title` flag has always set the window class not title. The `title` variable here is exactly what can be overwritten using the `--title` command line flag. What exactly do you mean by 'full solution' beyond what is currently happening @nixpulvis?

nixpulvis replied:
wait... `--title` sets the class? That seems very confusing. So I guess the "full solution" I see now is:

1. Re-purpose `--title` flag` to actual set the title, and prevent updates to running program names.
2. Create a `--class` flag that does what `--title` does today.

This will break peoples configs who are depending on this behavior, but this is weird enough (at least to me) to be worth it.

Tezkerek replied:
Well, the configs are already broken, because of changes to `winit`'s behaviour. And `--title` is also how termite does it, so I thought it would be fitting.

Also, [here](https://tronche.com/gui/x/xlib/ICC/client-to-window-manager/wm-class.html) it says that the first part of `WM_CLASS` is the application **name**, so if we don't use `--title`, we should use something like `--name` or `--classname`

nixpulvis replied:
@Tezkerek how termite does what? I'm pretty sure what I described is how termite works. I have no idea where the notion of a title came from however, that link makes it seem, indeed, like we should only have a `--name` and `--class` flag, which I'm just as happy with.

Tezkerek replied:
Yeah, I misremembered, you're right.

chrisduerr replied:
I'm not entirely sure what the original design idea behind this was since I haven't written it. But setting both title and class differently would be fine.

However, setting the name doesn't make much sense. I don't think the application name is used in a lot of places and instead we should differentiate between title and class. Only the class itself is mentioned in the URL I (and @Tezkerek) have linked. The title uses `WM_NAME` and is currently already properly set afaik.

So the current `--title` option sets only the title (displayed in the bar above the window in some WMs). In this PR this is already set as window class, it would be possible to change this to use `--class` instead. Then `--title` would only have an effect if dynamic window title is disabled in the config (which seems fine).

nixpulvis replied:
If I was understanding this correctly, the *name* is the same as the *title*. And *class* is the attribute that should be never changing (although possible to set via `--class`). The *title* Would then be (by default) the current running program, and possible to lock to a single value with `--title`.

nixpulvis replied:
Ok I also struggled with the [linked](https://tronche.com/gui/x/xlib/ICC/client-to-window-manager/wm-class.html) docs. But your [comment](https://github.com/jwilm/alacritty/issues/1350#issuecomment-394492246) cleared it up for me.

I think we're all saying the same thing at this point.

Tezkerek commented:
Hey @nixpulvis @chrisduerr, I've added the `--class` option. Instead of adding a second parameter to `Window::new` I removed the `title` parameter and replaced it with `cli::Options`.

I'm new to Rust, so please let me know if I did something wrong

chrisduerr commented on src/cli.rs:
```diff
@@ -81,6 +84,10 @@ impl Options {
+            .arg(Arg::with_name("class")
+                 .long("class")
+                 .default_value(DEFAULT_CLASS)
+                 .help("Defines WM_CLASS on supported platforms"))
```
It might be nice change the description up a bit. Something like `Defines window class on X11`. This is probably closer to what a user would expect and should make it obvious where this is supported.

It might also be possible to make use of platform-specific code here. Since MacOS will never support the window class, it might be nice to skip this option for MacOS. However in my experience splitting Clap into multiple functions is kind of a mess, so it's probably okay the way it is right now.

nixpulvis replied:
Skipping the option seems fine as long as we don't ever start using `--class` for something else on macOS. Which I can't imagine why we'd do.

chrisduerr replied:
Yeah, but it's probably easier to just leave it in for now, like that every user will get the same stuff.

chrisduerr commented on src/window.rs:
```diff
@@ -199,17 +200,19 @@ impl Window {
+
         let window_builder = WindowBuilder::new()
-            .with_title(title)
+            .with_title(title.to_owned())
```
You can actually use `&*options.title` here instead of defining a temporary variable.

chrisduerr commented on src/cli.rs:
```diff
@@ -81,6 +84,10 @@ impl Options {
                 .help("Defines the window title"))
+            .arg(Arg::with_name("class")
+                 .long("class")
+                 .default_value(DEFAULT_CLASS)
```
Sorry that I have to nitpick again, but I don't think `-c` is currently used by anything. So adding the `.short("c")` similar to the `.short("t")` for title would probably be a good addition.

Tezkerek replied:
It's not currently used, but I'd imagine the `--class` flag will likely be used rarely and mostly in scripts, so there's no harm in writing the long option. And in case a more important flag gets added it won't break anything.

chrisduerr replied:
Yeah I had a similar train of thought, especially because class is only supported on one platform.

It's probably a better idea to use `-c` for the already existing `--config-file`, thanks for speaking up. :)

sentriz commented:
thanks for this

nixpulvis commented:
Just doing some digging into this branch. First by running `xprop` on a few terminals.

### Without Any Arguments

```sh
# Alacritty
WM_CLASS(STRING) = "Alacritty", "Alacritty"
WM_NAME(STRING) = "xprop  /home/nixpulvis/Code/alacritty"

# Termite
WM_CLASS(STRING) = "termite", "Termite"
WM_NAME(STRING) = "xprop  /home/nixpulvis/Code/alacritty"

# urxvt
WM_CLASS(STRING) = "urxvt", "URxvt"
WM_NAME(STRING) = "xprop  /home/nixpulvis/Code/alacritty"
```

### With a `--class bar` Argument

```
WM_CLASS(STRING) = "bar", "Alacritty"
WM_CLASS(STRING) = "termite", "bar"
# urxvt doesn't report anything.
```
### With a `--title foo` Argument

```sh
# Alacritty doesn't report anything.
WM_ICON_NAME(STRING) = "foo"
# urxvt doesn't report anything.
```
So based on this, and the fact that I can't see the titles in my WM, I'd say we still need to lock the title/name attribute when passed a `--title` flag.

Also I find it interesting that we are the only terminal (out of three) to implement `--class` in a somewhat standard way. Termite updates the wrong of two strings, and urxvt doesn't support it.

nixpulvis commented:
I have the needed fixes implemented here: https://github.com/nixpulvis/alacritty/tree/wm-class-title-1351-1350.

chrisduerr commented:
@nixpulvis I'm sorry I wasn't able to respond in IRC, but I believe this option is already implemented:
https://github.com/jwilm/alacritty/blob/master/alacritty.yml#L222

chrisduerr commented:
I don't think the flag should override a config setting without giving the option to get back the default configuration setting. Otherwise it would be impossible to set the title at startup, but still have a dynamic window title.

jwilm commented:
In my experience, the title is almost always overwritten immediately when starting a shell. From my perspective, it probably wouldn't hurt to override the `dynamic_title` attribute when passing `--title`.

chrisduerr commented:
Yeah it would probably be immediately overwritten, but I thought it might still work with some window managers that move things based on title. So every window with title 'xyz' is moved, then it doesn't matter when it changes title immediately after.

However having to create an extra config file just to start up alacritty with a fixed title for the entire runtime seems pretty pointless to me, so maybe the best choice is to just say "use the class, not the title for that" and then add the change @nixpulvis has proposed.

However, I wouldn't add a separate config option but just automatically disable the `dynamic_title` setting whenever the command line flag is specified, that should do the trick.

@nixpulvis I'd like to know what you think about this since you've introduced the idea, does my proposal make sense to you? Also I'd probably just merge this and then just review your changes in a separate PR if you're fine with that. Otherwise you'd have to send a PR to @Tezkerek and that might just unnecessarily delay the process.

nixpulvis commented:
I'll have to test it, but I think the idea of just overwriting the config value `dynamic_title` will accomplish the same result as my larger change. As long as passing `--title` locks the title of the window this is good. The only potential edge case (the need for the `occurrences_of` addition) is that a flag of `--title` should disable `dynamic_title` but get the default value, and `--title ""` should give an empty title with `dynamic_title` also disabled.

chrisduerr commented:
@Tezkerek Thanks again for sending this PR, great work! :+1:
